File: src/app/html.ts
import template from './template.html?raw';

/**
 * The self-contained HTML exporter (README, PLAN §7.5): ONE file — the
 * compiled bundle + the figure JSON inlined into the template. Opening it
 * IS the instrument: full-viewport, live pan/zoom, debounced re-fit on
 * resize, hover-corner SVG / 4× PNG downloads (user rulings 2026-07-10).
 *
 * The template (`template.html`) is the single source of truth for the
 * page: the Python side vendors THE SAME FILE and performs THE SAME three
 * replacements — title text, the quoted figure token, the bundle comment
 * token. Replacements use function form (a `$` in the bundle or the JSON
 * must stay literal), and every `<` in the JSON is escaped to `\u003c`
 * so a title like `</script>…` can never break out of the script element.
 */

const escapeHtml = (s: string): string =>
  s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');

export function selfContainedHtml(figure: unknown, viewerJs: string): string {
  const title =
    typeof figure === 'object' &&
    figure !== null &&
    'title' in figure &&
    typeof (figure as { title: unknown }).title === 'string'
      ? (figure as { title: string }).title
      : 'coxeter-viz';
  const json = (JSON.stringify(figure) ?? 'null').replace(/</g, '\\u003c');
  return template
    .replace('__COXETER_VIZ_TITLE__', () => escapeHtml(title))
    .replace('"__COXETER_VIZ_FIGURE__"', () => json)
    .replace('/*__COXETER_VIZ_BUNDLE__*/', () => viewerJs);
}

File: python/src/coxeter_groups/_html.py
"""save('.html'): the three template replacements, mirroring the engine.

The vendored ``_static/template.html`` is the single source of truth for
the page; this module performs the SAME three replacements as the
TypeScript ``selfContainedHtml`` (src/app/html.ts) and the repo's
``scripts/make-samples.mjs``: the title text, the quoted figure token,
the bundle comment token. Every ``<`` in the JSON becomes ``\\u003c`` so
a hostile title can never break out of the script element. Python's
``str.replace`` has no ``$`` semantics, so the bundle text passes through
literally by construction (pinned by test anyway).
"""

from __future__ import annotations

import json
from importlib import resources
from typing import Any


def _static(name: str) -> str:
    ref = resources.files("coxeter_groups") / "_static" / name
    if not ref.is_file():
        from .figure import CoxeterVizError

        raise CoxeterVizError(
            f"the vendored engine file _static/{name} is missing — this is a development "
            "install without the bundle; run `npm run build:bundle` in the repo root."
        )
    return ref.read_text(encoding="utf-8")


def _escape_html(s: str) -> str:
    return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def page(document: dict[str, Any]) -> str:
    """The self-contained HTML page: template + bundle + inlined figure JSON."""
    template = _static("template.html")
    viewer = _static("viewer.js")
    title = document.get("title") or "coxeter-groups"
    figure_json = json.dumps(document).replace("<", "\\u003c")
    return (
        template.replace("__COXETER_VIZ_TITLE__", _escape_html(title))
        .replace('"__COXETER_VIZ_FIGURE__"', figure_json)
        .replace("/*__COXETER_VIZ_BUNDLE__*/", viewer)
    )
